caas/ecs/application.go

```diff
@@ -0,0 +1,637 @@
+				Autoprovision: aws.Bool(true),
+				Driver:        aws.String(ebsCfg.driver), // TODO: fs.Attributes["Driver"] ?????
+				Labels:        a.labels(),                // TODO: merge with fs.ResourceTags !!!
+				DriverOpts: map[string]*string{
+					"volumetype": aws.String(ebsCfg.volumeType),                    // TODO!!!
+					"size":       aws.String(strconv.FormatUint(fs.Size/1024, 10)), // unit of size here should be `Gi`
+				},
+			},
+		}
+		vols = append(vols, vol)
+
+		readOnly := false
+		if fs.Attachment != nil {
+			readOnly = fs.Attachment.ReadOnly
+		}
+		mounts = append(mounts, &ecs.MountPoint{
+			ContainerPath: aws.String(getMountPathForFilesystem(
+				idx, a.name, fs,
+			)),
+			SourceVolume: vol.Name,
+			ReadOnly:     aws.Bool(readOnly),
+		})
+	}
+	return vols, mounts, nil
+}
+
+func (a *app) applicationTaskDefinition(config caas.ApplicationConfig) (*ecs.RegisterTaskDefinitionInput, error) {
+
+	var containerNames []string
+	var containers []caas.ContainerConfig
+	for _, v := range config.Containers {
+		containerNames = append(containerNames, v.Name)
+		containers = append(containers, v)
+	}
+	sort.Strings(containerNames)
+	sort.Slice(containers, func(i, j int) bool {
+		return containers[i].Name < containers[j].Name
+	})
+
+	volumes, volumeMounts, err := a.handleFileSystems(config.Filesystems)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	input := &ecs.RegisterTaskDefinitionInput{
+		Family:      aws.String(a.name),
+		TaskRoleArn: aws.String(""),
+		ContainerDefinitions: []*ecs.ContainerDefinition{
+			// init container
+			{
+				Name:             aws.String("charm-init"),
+				Image:            aws.String(config.AgentImagePath),
+				WorkingDirectory: aws.String(jujuDataDir),
+				Cpu:              aws.Int64(10),
+				Memory:           aws.Int64(512),
+				Essential:        aws.Bool(false),
+				EntryPoint:       strPtrSlice("/opt/k8sagent"),
+				Command: strPtrSlice(
+					"init",
+					"--data-dir",
+					jujuDataDir,
+					"--bin-dir",
+					"/charm/bin",
+				),
+				Environment: []*ecs.KeyValuePair{
+					{
+						Name:  aws.String("JUJU_CONTAINER_NAMES"),
+						Value: aws.String(strings.Join(containerNames, ",")),
+					},
+					{
+						Name:  aws.String("JUJU_K8S_POD_NAME"),
+						Value: aws.String("cockroachdb-0"),
+					},
+					{
+						Name:  aws.String("JUJU_K8S_POD_UUID"),
+						Value: aws.String("c83b286e-8f45-4dbf-b2a6-0c393d93bd6a"),
+					},
+					// appSecret
+					{
+						Name:  aws.String("JUJU_K8S_APPLICATION"),
+						Value: aws.String(a.name),
+					},
+					{
+						Name:  aws.String("JUJU_K8S_MODEL"),
+						Value: aws.String(a.modelUUID),
+					},
+					{
+						Name:  aws.String("JUJU_K8S_APPLICATION_PASSWORD"),
+						Value: aws.String(config.IntroductionSecret),
+					},
+					{
+						Name:  aws.String("JUJU_K8S_CONTROLLER_ADDRESSES"),
+						Value: aws.String(config.ControllerAddresses),
+					},
+					{
+						Name:  aws.String("JUJU_K8S_CONTROLLER_CA_CERT"),
```

Comment:
So does this actually work as-is when you use the API? Does the SDK transparently encode the KV pairs when calling to the ECS API? I had issues when using the cli with a json document.

Reply:
It works as expected.

Reply:
It works as expected.

Reply:
Nice!